frontend/src/components/widgets/Checkbox/Checkbox.tsx

```diff
@@ -31,45 +31,46 @@ interface Props {
 interface State {
   /**
    * The value specified by the user via the UI. If the user didn't touch this
-   * widget's UI, it's undefined.
+   * widget's UI, the default value is used.
    */
-  value?: boolean;
+  value: boolean;
 }
 
 class Checkbox extends React.PureComponent<Props, State> {
-  public state: State = {}
+  public state: State = {
+    value: this.props.element.get('default'),
+  }
 
-  /**
-   * Return the user-entered value, or the widget's default value
-   * if the user hasn't interacted with it yet.
-   */
-  private get valueOrDefault(): boolean {
-    if (this.state.value === undefined) {
-      return this.props.element.get('value') as boolean
-    } else {
-      return this.state.value
+  public componentDidUpdate = (prevProps: Props): void => {
```

Comment:
As usual, React's lifecycle methods make worry :wink: 

Let me think about this some more... I'd like to see if there's an alternative to using `componentDidUpdate`

Reply:
Ok, wouldn't something like this work?

```
render() {
  // ...
  const out = <div> {/* ... */} </div>

  // This must be the last thing before returning.
  this.renderedElement = this.props.element
  return out
}

getValueOrDefault() {
  if (this.props.element == this.renderedElement) {
    return this.state.value
  } else {
    return this.props.element
  }
}
```

Reply:
We should definitely switch to Elm. There're no lifecycle hooks there 😄. You'll love it.

> Also, componentDidUpdate has a quirk: it doesn't get called for the initial render.

It doesn't have to. We know that the initial render should always happen with `this.props.element.get('default')`.

> Not to mention that the main purpose of this lifecycle method is to give you access to the DOM, which is not what we're doing here.

IMHO, it's one of its purposes - not the main. It's also used to do network requests, and a bunch of other stuff, including `setState` - which is exactly what we need here.

It's a nice solution, but it's not "Reacty". Moreover, to make it work we need to send a ForwardMessage when we use the default value (but we can't do it in render).

Reply:
Let's chat today? I just want to make sure we follow a clear pattern that we can describe in our "CONTRIBUTING" wiki page for everyone to follow.

Reply:
Ok, I just re-reviewed this with some more info in my head from the discussion over VC today, and I think the way to go is what @arraydude proposed: just move this logic into `getDerivedStateFromProps(newProps, state)`

Reply:
I used React keys solution as discussed.

Source: https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key